File: scripts/provider-fixtures.mjs
// SPDX-License-Identifier: Apache-2.0
import {readFileSync} from 'node:fs';
import {canonicalJson} from '@psp-cdl/core';
import {BufferedLlmLoop,createOpenAIChatProvider,OPENAI_CHAT_REVISION} from '@psp-cdl/llmproxy';
import {fixture as loopFixture} from './llm-fixtures.mjs';
export const suite=JSON.parse(readFileSync(new URL('../conformance/vectors/llm/openai-chat-0.1.json',import.meta.url),'utf8'));

export async function runCase(c) {
  const s=c.settings,requests=[],outputs=[],codes=[];
  let f,clock=1000,cancelled=!!s.preCancel,provider;
  const base=()=>f?.baseFlags;
  const effect=()=>{
    if(s.effect==='cancel'){cancelled=true;if(f)base().cancelled=true;}
    if(s.effect==='deadline'){clock=1800;if(f)base().now=1800;}
    if(s.effect==='expiry'&&f)base().now=1700;
    if(s.effect==='drift'&&f)base().drift=true;
  };
  try {
    if(c.scope==='loop')f=await loopFixture(s.loop??{});
    const transport=(body,_signal)=>{
      const request=JSON.parse(Buffer.from(body).toString('utf8'));requests.push(request);
      if(f)f.requests.push(request);
      effect();
      if(s.throwTransport)throw new Error('PRIVATE_BACKEND_EXCEPTION SYNTHETIC_KEY');
      const responses=s.responses??(f?[suite.tool,suite.final]:[suite.final]);
      const raw=s.rawBody??canonicalJson(responses[Math.min(requests.length-1,responses.length-1)]);
      return {status:s.status??200,contentType:s.contentType??'application/json',body:s.invalidUtf8?new Uint8Array([255]):Buffer.from(raw,'utf8')};
    };
    const config={mode:'offline',now:()=>f?f.loopHost.now():clock,complete:true,
      sources:[{id:'provider',capabilities:s.loop?.providerTraining?['used-for-model-training']:[]}],
      limits:{...suite.limits,...s.limits},transport,...s.config};
    if(config.mode==='live')delete config.transport;
    provider=createOpenAIChatProvider(config);
    if(f) {
      const snapshot=f.loopHost.snapshot;
      f.loopHost.snapshot=(...args)=>({...snapshot(...args),providerId:provider.id,providerRevision:s.loop?.wrongProvider?'wrong':OPENAI_CHAT_REVISION});
      const loop=new BufferedLlmLoop(f.store,f.gate,f.loopHost,provider);
      const result=await loop.run(s.loop?.token??'test-owner',f.session.sessionId,{message:'Read synthetic data.'},f.options);
      outputs.push(result);codes.push('OK');
    } else {
      for(let attempt=0;attempt<(s.attempts??1);attempt++) {
        try {outputs.push(await provider.invoke(structuredClone(s.request??suite.request),{deadline:s.deadline??1800,cancelled:()=>cancelled}));codes.push('OK');}
        catch(e){codes.push(e.code??'UNEXPECTED_ERROR');}
      }
    }
  }catch(e){codes.push(e.code??'UNEXPECTED_ERROR');}
  finally{f?.close();}
  return {code:codes.at(-1),codes,transportCalls:requests.length,toolCalls:f?.stats().toolCalls??0,released:outputs.length,requests,outputs};
}

File: scripts/provider_fixtures.py
# SPDX-License-Identifier: Apache-2.0
import json
from copy import deepcopy
from pathlib import Path
from psp_cdl_core import canonical_json
from psp_cdl_llmproxy import BufferedLlmLoop, create_openai_chat_provider, OPENAI_CHAT_REVISION
from llm_fixtures import Fixture

SUITE = json.loads((Path(__file__).resolve().parents[1]/"conformance/vectors/llm/openai-chat-0.1.json").read_text(encoding="utf-8"))


def run_case(case):
    settings, requests, outputs, codes = case["settings"], [], [], []
    fixture = None
    flags = {"now": 1000, "cancelled": bool(settings.get("preCancel"))}

    def transport(body, _stop):
        request = json.loads(body)
        requests.append(request)
        if fixture: fixture.requests.append(request)
        effect = settings.get("effect")
        if effect == "cancel":
            flags["cancelled"] = True
            if fixture: fixture.base.flags["cancelled"] = True
        if effect == "deadline":
            flags["now"] = 1800
            if fixture: fixture.base.flags["now"] = 1800
        if effect == "expiry" and fixture: fixture.base.flags["now"] = 1700
        if effect == "drift" and fixture: fixture.base.flags["drift"] = True
        if settings.get("throwTransport"): raise RuntimeError("PRIVATE_BACKEND_EXCEPTION SYNTHETIC_KEY")
        responses = settings.get("responses", [SUITE["tool"], SUITE["final"]] if fixture else [SUITE["final"]])
        raw = settings.get("rawBody", canonical_json(responses[min(len(requests)-1, len(responses)-1)]))
        return {"status": settings.get("status", 200), "contentType": settings.get("contentType", "application/json"),
                "body": b"\xff" if settings.get("invalidUtf8") else raw.encode("utf-8")}

    try:
        if case["scope"] == "loop": fixture = Fixture(deepcopy(settings.get("loop", {})))
        config = {"mode": "offline", "now": lambda: fixture.now() if fixture else flags["now"], "complete": True,
                  "sources": [{"id": "provider", "capabilities": ["used-for-model-training"] if settings.get("loop", {}).get("providerTraining") else []}],
                  "limits": {**SUITE["limits"], **settings.get("limits", {})}, "transport": transport, **settings.get("config", {})}
        if config["mode"] == "live": config.pop("transport")
        provider = create_openai_chat_provider(config)
        if fixture:
            snapshot = fixture.snapshot
            fixture.snapshot = lambda *args: {**snapshot(*args), "providerId": provider["id"], "providerRevision": "wrong" if settings.get("loop", {}).get("wrongProvider") else OPENAI_CHAT_REVISION}
            loop = BufferedLlmLoop(fixture.base.store, fixture.base.gate, fixture, provider)
            outputs.append(loop.run(settings.get("loop", {}).get("token", "test-owner"), fixture.base.session["sessionId"], {"message": "Read synthetic data."}, fixture.options))
            codes.append("OK")
        else:
            for _ in range(settings.get("attempts", 1)):
                try:
                    outputs.append(provider["invoke"](deepcopy(settings.get("request", SUITE["request"])), {"deadline": settings.get("deadline", 1800), "cancelled": lambda: flags["cancelled"]}))
                    codes.append("OK")
                except Exception as exc: codes.append(getattr(exc, "code", "UNEXPECTED_ERROR"))
    except Exception as exc: codes.append(getattr(exc, "code", "UNEXPECTED_ERROR"))
    finally:
        if fixture: fixture.close()
    return {"code": codes[-1], "codes": codes, "transportCalls": len(requests), "toolCalls": fixture.base.calls if fixture else 0,
            "released": len(outputs), "requests": requests, "outputs": outputs}
